File: CPP/ByRound/1781/1781_G_Diverse_Coloring.cpp
// Problem: CF 1781 G - Diverse Coloring
// https://codeforces.com/contest/1781/problem/G

/*
 * Problem: G. Diverse Coloring
 * 
 * Purpose: This code solves the problem of finding the minimum disbalance of a diverse coloring
 *          for a rooted binary tree that is built incrementally. A diverse coloring ensures
 *          that every vertex has at least one neighbor (parent or child) of the opposite color.
 *          The disbalance is defined as the absolute difference between the number of white
 *          and blue vertices.
 * 
 * Approach:
 * - For each step of adding a vertex, we compute the minimum disbalance for the current tree.
 * - We use a greedy method and bitmasking for small remaining components.
 * - The tree is represented with adjacency lists, and we process leaves first.
 * 
 * Algorithms/Techniques:
 * - Greedy algorithm for coloring
 * - BFS/DFS for traversal
 * - Bitmasking for small components
 * - Priority queue for leaf nodes
 * 
 * Time Complexity: O(n * 2^k) where n is the number of vertices and k is the size of the small component (k <= 10)
 * Space Complexity: O(n) for storing the tree and auxiliary arrays
 */

#include <algorithm>
#include <iostream>
#include <queue>
#include <utility>
#include <vector>

using namespace std;

int n, a[200010], p[200010], d[200010], ans[200010], vis[200010];
vector<int> v[200010];

int main() {
  int T;
  cin >> T;
  while (T--) {
    priority_queue<pair<int, int>> q;
    int cnt = 0, tot;
    cin >> n;
    for (int i = 1; i <= n; i++)
      vis[i] = 0, v[i].clear();
    for (int i = 2; i <= n; i++)
      cin >> p[i];
    // Print initial disbalance values based on specific conditions
    for (int i = 2; i <= n; i++) {
      if (i == 4)
        cout << (p[3] == p[4]) * 2 << '\n';
      else if (i & 1)
        cout << 1 << '\n';
      else
        cout << 0 << '\n';
    }
    // Build tree structure
    for (int i = 2; i <= n; i++)
      v[p[i]].push_back(i), d[i] = d[p[i]] + 1;
    // Initialize queue with leaf nodes
    for (int i = 2; i <= n; i++)
      if (!v[i].size())
        q.push(make_pair(d[i], i));
    tot = n;
    // Process the tree by removing leaf nodes in a greedy fashion
    while (tot > 10) {
      int x = q.top().second;
      q.pop();
      if (vis[x])
        continue;
      if (v[p[x]].size() == 1) {
        ans[p[x]] = 1;
        ans[x] = 0;
        vis[x] = vis[p[x]] = 1;
        tot -= 2;
        int y = p[p[x]];
        v[y].erase(find(v[y].begin(), v[y].end(), p[x]));
        if (!v[y].size())
          q.push(make_pair(d[y], y));
      } else {
        a[++cnt] = p[x];
        vis[v[p[x]][0]] = vis[v[p[x]][1]] = vis[p[x]] = 1;
        tot -= 3;
        int y = p[p[x]];
        v[y].erase(find(v[y].begin(), v[y].end(), p[x]));
        if (!v[y].size())
          q.push(make_pair(d[y], y));
      }
    }
    // Handle the remaining small component with bitmasking
    vector<int> b;
    for (int i = 1; i <= n; i++)
      if (!vis[i])
        b.push_back(i);
    int anss = 1e9, id;
    for (int i = 0; i < (1 << tot); i++) {
      int k = 0;
      for (int j = 0; j < tot; j++) {
        ans[b[j]] = (i >> j) & 1;
        if (i & (1 << j))
          k++;
        else
          k--;
      }
      if (k < 0)
        continue;
      int f = 1;
      for (int j = 0; j < tot; j++) {
        int ff = 0;
        if (p[b[j]] && ans[p[b[j]]] != ans[b[j]])
          ff = 1;
        for (auto y : v[b[j]])
          if (ans[b[j]] != ans[y])
            ff = 1;
        f &= ff;
      }
      if (f && k < anss)
        anss = k, id = i;
    }
    for (int j = 0; j < tot; j++)
      ans[b[j]] = (id >> j) & 1;
    // Final coloring update for the greedy components
    for (int i = 1; i <= cnt; i++) {
      if (anss > 0)
        ans[a[i]] = 1, ans[v[a[i]][0]] = ans[v[a[i]][1]] = 0, anss--;
      else
        ans[a[i]] = 0, ans[v[a[i]][0]] = ans[v[a[i]][1]] = 1, anss++;
    }
    // Print the final coloring result
    for (int i = 1; i <= n; i++)
      cout << (ans[i] ? 'w' : 'b');
    cout << endl;
  }
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1781/1781_G_Diverse_Coloring.py
# Problem: CF 1781 G - Diverse Coloring
# https://codeforces.com/contest/1781/problem/G

"""
Code Purpose:
This code solves the "G. Diverse Coloring" problem where we need to find the minimum disbalance 
in a binary tree's coloring after each vertex addition. The disbalance is defined as the absolute 
difference between the number of white and blue vertices. Each vertex must have at least one 
neighbor with a different color (diverse coloring).

Algorithms/Techniques:
- Tree construction from parent information
- Greedy approach based on vertex degrees to minimize disbalance
- Dynamic programming/assignment of colors based on degree ordering
- Special handling for edge cases like 4-node trees and specific degree configurations

Time Complexity: O(n log n) due to sorting of vertices by degree
Space Complexity: O(n) for storing tree structure and auxiliary arrays
"""

import sys


def read():
    s = 0
    w = 1
    ch = sys.stdin.read(1)
    while ch < "0" or ch > "9":
        if ch == "-":
            w = -1
        ch = sys.stdin.read(1)
    while "0" <= ch <= "9":
        s = s * 10 + ord(ch) - ord("0")
        ch = sys.stdin.read(1)
    return s * w


fa = [0] * 1000003  # Parent of each vertex
deg = [0] * 1000003  # Degree of each vertex (number of children)
ls = [0] * 1000003   # Left child of each vertex  
rs = [0] * 1000003   # Right child of each vertex
flg = [0] * 1000003  # Flag to mark if vertex is processed
ans = [0] * 1000003  # Color assignment (0 = blue, 1 = white) for each vertex


def main():
    T = read()
    for _ in range(T):
        n = read()

        # Initialize arrays for current test case
        for i in range(1, n + 1):
            deg[i] = ls[i] = rs[i] = 0
            flg[i] = ans[i] = 0

        # Build tree structure from parent information
        for i in range(2, n + 1):
            fa[i] = read()
            if ls[fa[i]]:  # If left child exists, set right child
                rs[fa[i]] = i
            else:  # Otherwise set left child
                ls[fa[i]] = i

        # Print intermediate disbalance values
        # This seems like placeholder code or debugging output
        for i in range(2, n + 1):
            if i == 4:
                print((fa[3] == fa[4]) << 1)
            else:
                print(i & 1)

        # Special case handling for n=4
        if n == 4 and fa[3] == fa[4]:
            print("wbww")
            continue

        # Calculate degrees of all vertices
        for i in range(n, 1, -1):  # Process from n down to 2
            if not flg[i]:  # If not yet processed
                deg[fa[i]] += 1  # Increment parent's degree
                flg[fa[i]] = 1   # Mark parent as processed

        # Process root (vertex 1) if not processed
        if not flg[1]:
            if deg[2] == 2 and rs[1] and deg[rs[1]] == 1:
                # Special case for degree configuration
                deg[1] = deg[rs[1]] + 1
                deg[rs[1]] = 0
                flg[rs[1]] = 0
                flg[1] = 1
                ans[1] = 1
            else:
                deg[1] = deg[2] + 1
                deg[2] = 0
                flg[2] = 0
                flg[1] = 1
                ans[1] = 1

                # Additional handling for specific degree configurations
                if deg[1] == 3 and not rs[1]:
                    ok = 0
                    for i in range(2, n + 1):
                        if deg[i] == 2:
                            ok = 1
                    if not ok:
                        x = y = 0
                        # Find children of left child of vertex 2
                        if ls[ls[2]]:
                            x, y = ls[2], ls[ls[2]]
                        if rs[ls[2]]:
                            x, y = ls[2], rs[ls[2]]
                        if ls[rs[2]]:
                            x, y = rs[2], ls[rs[2]]
                        if rs[rs[2]]:
                            x, y = rs[2], rs[rs[2]]

                        deg[1] -= 1
                        flg[x] = 1
                        flg[y] = 0
                        ans[x] = 1
                        deg[x] = deg[y] + 1
                        deg[y] = 0

        # Collect vertices with degree > 1 for processing
        vec = []
        for i in range(1, n + 1):
            if deg[i] > 1:
                vec.append((deg[i] - 1, i))

        # Sort vertices by degree (descending)
        vec.sort(reverse=True)

        # Greedily assign colors to minimize disbalance
        total_sum = 0
        for x, y in vec:
            if total_sum > 0:
                total_sum -= x
                ans[y] ^= 1  # Toggle color
            else:
                total_sum += x

        # Assert correctness of final sum (should match parity of n)
        assert total_sum == (n & 1)

        # Assign colors to remaining unassigned vertices (based on parent)
        for i in range(2, n + 1):
            if not flg[i]:
                ans[i] = ans[fa[i]] ^ 1  # Inherit opposite color from parent

        # Format output string and validate result
        S = sum(ans[1 : n + 1])  # Count white vertices
        for i in range(1, n + 1):
            print("w" if ans[i] else "b", end="")  # Output coloring

        real = abs(S * 2 - n)  # Calculate actual disbalance
        assert real == total_sum  # Validate

        print("")


if __name__ == "__main__":
    main()


# https://github.com/VaHiX/CodeForces/